### PR TITLE
local alert resync

### PR DESCRIPTION
**Pull request by fspmarshall**
Previously, upgrade alerts stuck around until the next time teleport scraped the GitHub releases page, which only happens a couple times per day.  This could cause upgrade suggestions to continue to be displayed for many hours after upgrading, which is very confusing.

With this PR the auth server now adds a label to upgrade alerts indicating what version it observed that prompted the alert to be generated.  We then periodically use these labels to recalculate what alerts to display.  Because we do a full recalculation rather than just deleting alerts that seem outdated, this means alerts can also be "downgraded". Ex:  if the auth server has been updated but some of the nodes haven't been, the message automatically downgrades from the "please update your cluster" case to the "some instances appear outdated" case. 

Because upgrade alerts are now frequently recalculated, they have a much shorter TTL (30m, previously was 24h).

This PR does not change the frequency with which discovery of *new* releases is attempted.

This is a temporary workaround to hold us over until we have a proper system for caching release targets  (the `version-directive` resource from the upgrades RFD) .  Once that is in place, we'll be able to recalculate alerts from canonical source material.

This PR also limits notifications in cloud to only inform users of releases less than or equal to the current auth version.  Since it is our responsibility to upgrade cloud auth and we have to do that before customers can upgrade their peripheral instances, alerts about newer versions are unnecessary noise.

### CODE REVIEW

**Comment by zmb3** [comment]
Do we need to check every 10 minutes? The version that's currently running can only change after a process restart, so can't we get away with a single check on startup?

File: lib/auth/auth.go [hunk 1]
```diff
@@ -563,13 +576,74 @@ func (a *Server) checkForReleases(ctx context.Context) {
...
+// re-pulling the releases page. Future version of teleport will cache actual full release
+// descriptions, rending this unnecessary.
+func (a *Server) visitCachedAlertVersions(ctx context.Context, visitor *vc.Visitor) error {
+	// reconstruct the target for the "latest stable" alert if it exists.
+	alert, err := a.getClusterAlert(ctx, releaseAlertID)
+	if err != nil && !trace.IsNotFound(err) {
+		return trace.Wrap(err)
 	}
+	if !trace.IsNotFound(err) {
+		if t := vc.NewTarget(alert.Metadata.Labels[verInUseLabel]); t.Ok() {
```
Comment (smallinsky) [inline lib/auth/auth.go:614]: Also the alert is accessed in case of err != nil

File: lib/auth/auth.go [hunk 2]
```diff
@@ -563,13 +576,74 @@ func (a *Server) checkForReleases(ctx context.Context) {
...
+func (a *Server) getClusterAlert(ctx context.Context, id string) (types.ClusterAlert, error) {
+	alerts, err := a.GetClusterAlerts(ctx, types.GetClusterAlertsRequest{
+		AlertID: id,
+	})
+	if err != nil || len(alerts) == 0 {
+		if err == nil {
+			err = trace.NotFound("cluster alert %q not found")
+		}
+		return types.ClusterAlert{}, trace.Wrap(err)
+	}
```
Comment (smallinsky) [inline lib/auth/auth.go:641]: nit: I think that sequential flow is more readable: 
```suggestion
	if err != nil {
		return types.ClusterAlert{}, trace.Wrap(err)
	}
	 if len(alerts) == 0 {
	      return types.ClusterAlert{}, trace.NotFound("cluster alert %q not found", id)
	 } 
```

File: lib/auth/auth.go [hunk 3]
```diff
@@ -563,13 +576,74 @@ func (a *Server) checkForReleases(ctx context.Context) {
...
+	return nil
+}
+
+func (a *Server) getClusterAlert(ctx context.Context, id string) (types.ClusterAlert, error) {
+	alerts, err := a.GetClusterAlerts(ctx, types.GetClusterAlertsRequest{
+		AlertID: id,
+	})
+	if err != nil || len(alerts) == 0 {
+		if err == nil {
+			err = trace.NotFound("cluster alert %q not found")
```
Comment (smallinsky) [inline lib/auth/auth.go:638]:
```suggestion
			err = trace.NotFound("cluster alert %q not found", id)
```

File: lib/auth/auth.go [hunk 4]
```diff
@@ -563,13 +576,74 @@ func (a *Server) checkForReleases(ctx context.Context) {
...
+// of existing alerts. This lets us "reevaluate" the alerts based on newer cluster state without
+// re-pulling the releases page. Future version of teleport will cache actual full release
+// descriptions, rending this unnecessary.
+func (a *Server) visitCachedAlertVersions(ctx context.Context, visitor *vc.Visitor) error {
+	// reconstruct the target for the "latest stable" alert if it exists.
+	alert, err := a.getClusterAlert(ctx, releaseAlertID)
+	if err != nil && !trace.IsNotFound(err) {
+		return trace.Wrap(err)
 	}
+	if !trace.IsNotFound(err) {
```
[the remark this answers is not in the document]
Comment (fspmarshall) [reply lib/auth/auth.go:613]: The preceding check is `if err != nil && !trace.IsNotFound(err) { ... }`, so when we get to this line, we either have `nil` or `NotFound`.  This check can be true in the `NotFound` case.  I'll change it to read `if err == nil { ... }`.  The two are equivalent in this case, but the latter is probably going to be less confusing.

File: lib/auth/auth.go (hunk 1 above)
Comment (fspmarshall) [reply lib/auth/auth.go:614]: The preceding check is `if err != nil && !trace.IsNotFound(err) { ... }`, so when we get to this line, we either have `nil` or `NotFound`.  I'm gonna change this check to read `if err == nil { ... }`.  The two are equivalent in this case, but it'll be less confusing.

**Comment by fspmarshall** [comment]
@zmb3 Things don't just change on process restart.  The alert also needs to change if people upgrade their non-auth instances.

**Review by rosstimothy — APPROVED** [review]

**Review by gabrielcorado — APPROVED** [review]

**Pull request merged**